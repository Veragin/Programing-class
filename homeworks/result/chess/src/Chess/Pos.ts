export const letters = ["A", "B", "C", "D", "E", "F", "G", "H"] as TY[];

type TPos = { x: TX; y: TY };
type TIndex = { i: number; j: number };

export class Pos {
    private _i: number;
    private _j: number;

    constructor(input: TPos | TIndex) {
        const inp = input as any;
        if (inp.x !== undefined) {
            this._i = inp.x - 1;
            this._j = letters.indexOf(inp.y);
        } else {
            this._i = inp.i;
            this._j = inp.j;
        }
    }

    get x() {
        return (this._i + 1) as TX;
    }
    get y() {
        return letters[this._j];
    }
    get i() {
        return this._i;
    }
    get j() {
        return this._j;
    }

    isEqual(pos: Pos) {
        return this.i === pos.i && this.j === pos.j;
    }

    export() {
        return [this._i, this._j];
    }
}
