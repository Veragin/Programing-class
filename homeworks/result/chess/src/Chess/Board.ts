import { BOARD_SIZE } from "./Const";
import { ChessPiece } from "./Pieces/ChessPiece";
import { Pos } from "./Pos";

export class Board {
    board: TBoard = [];

    constructor() {
        for (let i = 0; i < BOARD_SIZE; i++) {
            this.board.push([]);
            for (let j = 0; j < BOARD_SIZE; j++) {
                this.board[i].push(null);
            }
        }
    }

    set(piece: ChessPiece, pos: Pos) {
        try {
            const { i, j } = piece.findMyPosInBoard(this.board);
            this.board[i][j] = null;
        } catch {}

        this.board[pos.i][pos.j] = piece;
    }
}
