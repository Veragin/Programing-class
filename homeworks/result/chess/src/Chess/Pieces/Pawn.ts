import { Pos } from "../Pos";
import { ChessPiece } from "./ChessPiece";

export class Pawn extends ChessPiece {
    move(board: TBoard, myPos: Pos) {
        const res: Pos[] = [];
        const i = myPos.i;
        const j = myPos.j;

        const newI = this.player === "black" ? i - 1 : i + 1;

        if (board[newI][j] === null) {
            res.push(new Pos({ i: newI, j }));
        }

        let p = board[newI][j - 1];
        if (p instanceof ChessPiece && p.player !== this.player) {
            res.push(new Pos({ i: newI, j: j - 1 }));
        }

        p = board[newI][j + 1];
        if (p instanceof ChessPiece && p.player !== this.player) {
            res.push(new Pos({ i: newI, j: j + 1 }));
        }

        return res;
    }
}
