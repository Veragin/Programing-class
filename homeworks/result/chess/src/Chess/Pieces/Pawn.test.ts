import { Board } from "../Board";
import { Pos } from "../Pos";
import { Pawn } from "./Pawn";

const createData = () => {
    const board = new Board();
    const pawn = new Pawn("white");
    const pawnPos = new Pos({ x: 2, y: "B" });

    board.board[pawnPos.i][pawnPos.j] = pawn;

    return {
        board: board.board,
        pawn,
        pawnPos,
    };
};

describe("Pawn", () => {
    it("can move", () => {
        const { board, pawn, pawnPos } = createData();

        board[pawnPos.i + 1][pawnPos.j + 1] = new Pawn("black");

        const res = pawn.move(board, pawnPos);
        expect(res.length).toBe(2);

        const exp = res.map((pos) => pos.export());
        expect(exp).toMatchSnapshot();
    });

    it("cannt move if is blocked by figure", () => {
        const { board, pawn, pawnPos } = createData();

        board[pawnPos.i + 1][pawnPos.j] = new Pawn("black");

        const res1 = pawn.move(board, pawnPos);
        expect(res1.length).toBe(0);

        board[pawnPos.i + 1][pawnPos.j] = new Pawn("white");

        const res2 = pawn.move(board, pawnPos);
        expect(res2.length).toBe(0);
    });
});

export {};
