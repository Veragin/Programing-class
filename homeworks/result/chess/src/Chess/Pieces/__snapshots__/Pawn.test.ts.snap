// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pawn can move 1`] = `
Array [
  Array [
    2,
    1,
  ],
  Array [
    2,
    2,
  ],
]
`;
